chips/sam4l/src/pm.rs

```diff
@@ -530,61 +555,289 @@ const PM_REGS: StaticRef<PmRegisters> = unsafe { StaticRef::new(PM_BASE as *cons
+                flashcalw::FLASH_CONTROLLER.set_wait_state(0);
+                // Change the system clock to RCSYS
+                select_main_clock(MainClock::RCSYS);
             }
 
             SystemClockSource::DfllRc32kAt48MHz => {
+                // Configure and turn on DFLL at 48MHz
                 configure_48mhz_dfll();
-                self.system_frequency.set(48000000);
+                // Set Flash wait state to 1 for > 24MHz in PS2
+                flashcalw::FLASH_CONTROLLER.set_wait_state(1);
+                // Change the system clock to DFLL
+                select_main_clock(MainClock::DFLL);
             }
 
             SystemClockSource::ExternalOscillator {
                 frequency,
                 startup_mode,
             } => {
+                match self.system_clock_source.get() {
+                    // If the PLL is running (it uses OSC0 as a reference clock),
+                    // temporarily change the system clock to RCSYS to avoid buggy behavior
+                    SystemClockSource::PllExternalOscillatorAt48MHz { .. } => {
+                        select_main_clock(MainClock::RCSYS);
+                    }
+                    // Some peripherals (uart,spi) show buggy behavior if the system clock
+                    // is directly switched from OSC0 to DFLL - no explanation in documentation
+                    SystemClockSource::DfllRc32kAt48MHz => {
+                        select_main_clock(MainClock::RCSYS);
+                    }
+                    _ => {}
+                }
+                // Configure and turn on OSC0
                 configure_external_oscillator(frequency, startup_mode);
-                match frequency {
-                    OscillatorFrequency::Frequency16MHz => self.system_frequency.set(16000000),
-                };
+                // Set Flash wait state to 0 for <= 24MHz in PS2
+                flashcalw::FLASH_CONTROLLER.set_wait_state(0);
+                // Change the system clock to OSC0
+                select_main_clock(MainClock::OSC0);
             }
 
             SystemClockSource::PllExternalOscillatorAt48MHz {
                 frequency,
                 startup_mode,
             } => {
+                // Configure and turn on PLL at 48MHz
                 configure_external_oscillator_pll(frequency, startup_mode);
-                self.system_frequency.set(48000000);
+                // Set Flash wait state to 1 for > 24MHz in PS2
+                flashcalw::FLASH_CONTROLLER.set_wait_state(1);
+                // Change the system clock to PLL
+                select_main_clock(MainClock::PLL);
+            }
+
+            SystemClockSource::RC80M => {
+                // Configure and turn on RC80M
+                configure_80mhz_rc();
+
+                // If the 80MHz RC is used as the main clock source, it must be divided by
+                //  at least 2 before being used as CPU's clock source
+                unlock(0x00000004);
```

Comment:
Does this, and the other uses of `unlock` in this file, need to change to the `modify_no_read` pattern?